File: 17-merge-sorted-lists/solution.py
# Merge Two Sorted Linked Lists
# Time: O(m + n), Space: O(1)

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode(0)
        curr = dummy
        
        while l1 and l2:
            if l1.val <= l2.val:
                curr.next = l1
                l1 = l1.next
            else:
                curr.next = l2
                l2 = l2.next
            curr = curr.next
        
        curr.next = l1 if l1 else l2
        
        return dummy.next

# Test
if __name__ == "__main__":
    # Create lists: 1->2->4 and 1->3->4
    l1 = ListNode(1, ListNode(2, ListNode(4)))
    l2 = ListNode(1, ListNode(3, ListNode(4)))
    
    sol = Solution()
    merged = sol.mergeTwoLists(l1, l2)
    
    curr = merged
    result = []
    while curr:
        result.append(curr.val)
        curr = curr.next
    print(result)  # [1, 1, 2, 3, 4, 4]

File: 17-merge-sorted-lists/solution.cs
// Merge Two Sorted Linked Lists
// Time: O(m + n), Space: O(1)

using System;

public class ListNode {
    public int val;
    public ListNode next;
    
    public ListNode(int val = 0, ListNode next = null) {
        this.val = val;
        this.next = next;
    }
}

public class Solution {
    public ListNode MergeTwoLists(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(0);
        ListNode curr = dummy;
        
        while (l1 != null && l2 != null) {
            if (l1.val <= l2.val) {
                curr.next = l1;
                l1 = l1.next;
            } else {
                curr.next = l2;
                l2 = l2.next;
            }
            curr = curr.next;
        }
        
        curr.next = l1 ?? l2;
        
        return dummy.next;
    }
    
    public static void Main(string[] args) {
        ListNode l1 = new ListNode(1, new ListNode(2, new ListNode(4)));
        ListNode l2 = new ListNode(1, new ListNode(3, new ListNode(4)));
        
        Solution sol = new Solution();
        ListNode merged = sol.MergeTwoLists(l1, l2);
        
        while (merged != null) {
            Console.Write(merged.val + (merged.next != null ? " " : "\n"));
            merged = merged.next;
        }
    }
}
